pkg/gadgets/socket-collector/gadget.go

```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
 	params map[string]string) {
 
 	if trace.ObjectMeta.Namespace != gadgets.TRACE_DEFAULT_NAMESPACE {
-		gadgets.SetStatusError(trace, fmt.Sprintf("This gadget only accepts operations on traces in the %s namespace", gadgets.TRACE_DEFAULT_NAMESPACE))
+		gadgets.SetStatusError(trace, fmt.Sprintf("This gadget only accepts operations on traces in the %s namespace",
+			gadgets.TRACE_DEFAULT_NAMESPACE))
 		return
 	}
 
 	switch operation {
 	case "start":
-		var pid uint32 = 0
-
 		if trace.Spec.Filter == nil {
 			gadgets.SetStatusError(trace, "Missing filter")
 			return
 		}
-		if trace.Spec.Filter.Namespace == "" {
-			gadgets.SetStatusError(trace, "Invalid filter: missing namespace")
-			return
-		}
-		if trace.Spec.Filter.Podname == "" {
-			gadgets.SetStatusError(trace, "TODO: Filtering only by namespace is not currently supported")
-			return
-		}
-		if len(trace.Spec.Filter.Labels) != 0 {
-			gadgets.SetStatusError(trace, "TODO: Filtering by labels is not currently supported")
-			return
-		}
 		if trace.Spec.Filter.ContainerName != "" {
-			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!", trace.Spec.Gadget)
+			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!",
+				trace.Spec.Gadget)
 		}
 
-		pidsInPod := t.resolver.LookupPIDByPod(trace.Spec.Filter.Namespace, trace.Spec.Filter.Podname)
-
-		// All containers inside a pod share the same network namespace
-		// thus we can just take the first valid pid we found
-		for _, pid = range pidsInPod {
-			if pid != 0 {
-				break
-			}
+		selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
+		filteredContainers := t.resolver.GetContainersBySelector(selector)
+		if len(filteredContainers) == 0 {
+			gadgets.SetStatusError(trace, fmt.Sprintf("No container matches the requested filter: %+v",
+				*trace.Spec.Filter))
+			return
```

Comment:
Even if this test would make sense it's not the place to do it. This is the list of containers of **this** node, so if I have 10 nodes and I ask for the sockets of a given pod I'll get 9 error messages from the nodes where the pod it not scheduled.

```
$ kubectl get nodes -A
NAME           STATUS   ROLES                  AGE     VERSION
minikube       Ready    control-plane,master   10d     v1.21.2
minikube-m02   Ready    <none>                 3m44s   v1.21.2
$ kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
mypod   1/1     Running   3          10d
$ ./kubectl-gadget-linux-amd64 socket-collector -p mypod
WARN[0002] Error getting traces from node "minikube-m02": No container matches the requested filter: {Namespace:default Podname:mypod Labels:map[] ContainerName:}  args="[]" command="kubectl-gadget socket-collector"
NODE        NAMESPACE    POD      PROTOCOL    LOCAL           REMOTE       STATUS
minikube    default      mypod    TCP         0.0.0.0:5000    0.0.0.0:0    LISTEN

```

Reply:
I reproduced this situation in my environment. I agree that message should not be displayed. To do so, I modified the code of the socket-collector gadget to not finish the trace with an error when no container matches the requested filter, but instead, just print that message as a warning and put the trace as `Completed` with an empty output. Now, the output is clean:

```
$ ./kubectl-gadget-linux-amd64 socket-collector -A -p pod-only-on-node1
NODE     NAMESPACE      POD                  PROTOCOL    LOCAL                 REMOTE             STATUS
node1    kube-system    pod-only-on-node1    TCP         192.168.49.3:46468    10.96.0.1:443      ESTABLISHED
node1    kube-system    pod-only-on-node1    TCP         192.168.49.3:46470    10.96.0.1:443      ESTABLISHED
```

However, in the case of a non-existent pod, it will then show an empty list which, as we discussed before, could be misunderstood. Therefore, I modified the `utils/trace` in the `kubectl-gadget` side to not consider an empty output as a "successful node" (`successNodeCount`) when the `OutputMode` is different than `Stream`. Doing so, we get the following output for a non-existent pod:

```
$ ./kubectl-gadget-linux-amd64 socket-collector -A -p non-existent-pod
FATA[0000] Error getting traces from all nodes           args="[]" command="kubectl-gadget socket-collector"
```
I still would like to improve this to show the final user that the actual reason was that there were no matches with the requested filter. Right now, this info is only available in the logs of the IG running in the involved node. However, I would let that improvement for a future PR.

What do you think?

Reply:
I think it works. I put a comment regarding that implementation.

Reply:
I applied your suggestion.